### PR TITLE
An assortment of compiler and schema fixes

### PR DESCRIPTION
**Pull request by elprans**
This is a series of fixes and general cleanups that I needed to do as a prerequisite for the schema reflection rework.

Please review commits individually.

### CODE REVIEW

File: edb/edgeql/compiler/context.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 import enum
 import uuid
 
-from dataclasses import dataclass
+from dataclasses import dataclass, field as dc_field
```
Comment (1st1) [inline edb/edgeql/compiler/context.py:30]: Fix this to module import please.

File: edb/ir/ast.py [hunk 2]
```diff
@@ -667,7 +667,7 @@ class Stmt(Expr):
 class FilteredStmt(Stmt):
     __abstract_node__ = True
     where: Set
-    where_card: qltypes.SchemaCardinality
+    where_card: qltypes.Cardinality
```
Comment (1st1) [inline edb/ir/ast.py:671]: 🤦‍♂️

File: edb/edgeql/ast.py [hunk 3]
```diff
@@ -318,6 +318,7 @@ def from_python(cls, s: bytes) -> BytesConstant:
 
 class Parameter(Expr):
     name: str
+    optional: bool = True
```
Comment (1st1) [inline edb/edgeql/ast.py:321]: Are query arguments actually optional by default, e.g. would `fetchone('SELECT <int64>$0 +1', None)` work? Is that a good default?

```python
>>> c.fetchone('select <int64>$0 + 1', None)
Traceback (most recent call last):
  File "edgedb/protocol/protocol.pyx", line 409, in edgedb.protocol.protocol.SansIOProtocol._optimistic_execute
  File "edgedb/protocol/protocol.pyx", line 1089, in edgedb.protocol.protocol.SansIOProtocol.parse_data_messages
  File "edgedb/protocol/./codecs/./scalar.pyx", line 31, in edgedb.protocol.protocol.ScalarCodec.decode
  File "edgedb/pgproto/./codecs/int.pyx", line 118, in edgedb.pgproto.pgproto.int8_decode
  File "edgedb/pgproto/./frb.pxd", line 28, in edgedb.pgproto.pgproto.frb_read
  File "edgedb/pgproto/./frb.pyx", line 10, in edgedb.pgproto.pgproto.frb_check
AssertionError: insufficient data in buffer: requested 8 remaining 0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/yury/dev/venvs/edgedb/lib/python3.8/site-packages/edgedb/blocking_con.py", line 107, in fetchone
    return self._protocol.sync_execute_anonymous(
  File "edgedb/protocol/blocking_proto.pyx", line 91, in edgedb.protocol.blocking_proto.BlockingIOProtocol.sync_execute_anonymous
  File "edgedb/protocol/blocking_proto.pyx", line 77, in edgedb.protocol.blocking_proto.BlockingIOProtocol._iter_coroutine
  File "edgedb/protocol/protocol.pyx", line 555, in execute_anonymous
  File "edgedb/protocol/protocol.pyx", line 445, in _optimistic_execute
edgedb.errors.ClientError: unable to decode data to Python objects
```

🤯

File: edb/schema/referencing.py [hunk 4]
```diff
@@ -18,14 +18,16 @@
 
 
 from __future__ import annotations
-
-import hashlib
 from typing import *
 
-from edb.edgeql import ast as qlast
+import hashlib
```
Comment (1st1) [inline edb/schema/referencing.py:23]: Don't see how you use it in the diff.

Comment (1st1) [reply edb/schema/referencing.py:23]: Never mind.

File: edb/edgeql/ast.py (hunk 3 above)
Comment (elprans) [reply edb/edgeql/ast.py:321]: > Is that a good default?

I'm actually thinking we should make query arguments required by default.

Comment (1st1) [reply edb/edgeql/ast.py:321]: I always assumed they are, so this is news to me. ;(

**Pull request merged**